File: skills/code-architect/references/code-examples/go/memento.go
// Memento: capture an object's internal state so it can be restored later
// without violating its encapsulation.
package main

import "fmt"

// memento is an opaque snapshot; only the originator reads its contents.
type memento struct {
	state string
}

// originator owns the state and creates/restores snapshots.
type originator struct {
	state string
}

func (o *originator) Set(s string) { o.state = s }

func (o *originator) Save() memento { return memento{state: o.state} }

func (o *originator) Restore(m memento) { o.state = m.state }

// caretaker stores snapshots but never inspects them.
type caretaker struct {
	history []memento
}

func (c *caretaker) push(m memento) { c.history = append(c.history, m) }

func (c *caretaker) pop() memento {
	n := len(c.history) - 1
	m := c.history[n]
	c.history = c.history[:n]
	return m
}

func main() {
	o := &originator{}
	care := &caretaker{}

	o.Set("draft")
	care.push(o.Save())

	o.Set("edited")
	fmt.Println("current:", o.state)

	o.Restore(care.pop())
	fmt.Println("restored:", o.state)
}

File: skills/code-architect/references/code-examples/rust/memento.rs
// Memento — capture and externalize an object's internal state so it can be
// restored later, without violating encapsulation.

// Originator owns the state and produces/consumes mementos.
struct Editor {
    text: String,
    cursor: usize,
}

// Memento is an opaque snapshot to outsiders.
struct Snapshot {
    text: String,
    cursor: usize,
}

impl Editor {
    fn new() -> Self {
        Editor { text: String::new(), cursor: 0 }
    }

    fn type_text(&mut self, s: &str) {
        self.text.push_str(s);
        self.cursor = self.text.len();
    }

    fn save(&self) -> Snapshot {
        Snapshot { text: self.text.clone(), cursor: self.cursor }
    }

    fn restore(&mut self, snap: Snapshot) {
        self.text = snap.text;
        self.cursor = snap.cursor;
    }

    fn state(&self) -> String {
        format!("'{}' (cursor={})", self.text, self.cursor)
    }
}

fn main() {
    // Caretaker keeps mementos without inspecting their contents.
    let mut editor = Editor::new();
    let mut history: Vec<Snapshot> = Vec::new();

    editor.type_text("hello");
    history.push(editor.save());
    editor.type_text(" world");
    println!("current: {}", editor.state());

    editor.restore(history.pop().unwrap());
    println!("restored: {}", editor.state());
}
